infoschema/cache.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infoschema
+
+import (
+	"sort"
+	"sync"
+)
+
+// InfoCache handles information schema, including getting and setting.
+// The cache behavior, however, is transprent and under automatic management.
+// It only promised to cache the infoschema, if it is newer than all the cached.
+type InfoCache struct {
+	mu sync.RWMutex
+	// cache is sorted by SchemaVersion in descending order
+	cache []InfoSchema
+}
+
+// NewCache creates a new InfoCache.
+func NewCache(capcity int) *InfoCache {
+	return &InfoCache{cache: make([]InfoSchema, 0, capcity)}
+}
+
+// GetLatest gets the newest information schema.
+func (h *InfoCache) GetLatest() InfoSchema {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+	if len(h.cache) > 0 {
+		return h.cache[0]
+	}
+	return nil
+}
+
+// GetByVersion gets the information schema based on schemaVersion. Returns nil if it is not loaded.
+func (h *InfoCache) GetByVersion(version int64) InfoSchema {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+	i := sort.Search(len(h.cache), func(i int) bool {
+		return h.cache[i].SchemaMetaVersion() <= version
+	})
+	if i < len(h.cache) && h.cache[i].SchemaMetaVersion() == version {
+		return h.cache[i]
+	}
+	return nil
+}
+
+// Insert will **TRY** to insert the infoschema into the cache.
+// It only promised to cache the newest infoschema.
+func (h *InfoCache) Insert(is InfoSchema) {
```

Comment:
I suggest to return `bool` which indicates whether insert successfully or not.

Reply:
Done. There was a return value, but it is removed due to no usage.

Reply:
I think it might be helpful if we try to add a metrics to record the status for infoCache.

Reply:
Do you mean someting like hit rate or so?